src/checker.rs

```diff
@@ -1,76 +1,157 @@
 use std::error::Error;
 
 use reqwest::{Response, StatusCode};
+use tokio::time::Instant;
+use uuid::Uuid;
 
-#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
-pub enum CheckResult {
-    Success,
-    Failure(FailureReason),
-    Missed,
-}
+use crate::types::{
+    CheckResult, CheckStatus, CheckStatusReason, CheckStatusReasonType, RequestInfo, RequestType,
+};
 
-#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
-pub enum FailureReason {
-    Timeout,
-    DnsError(String),
-    Error(String),
-}
 /// Fetches the response from a URL.
 ///
 /// First attempts to fetch just the head, and if not supported falls back to fetching the entire body.
-async fn do_request(client: &reqwest::Client, url: &str) -> Result<Response, reqwest::Error> {
-    let response = client.head(url).send().await?;
-    match response.status() {
-        StatusCode::METHOD_NOT_ALLOWED => client.get(url).send().await,
-        _ => Ok(response),
+async fn do_request(
+    client: &reqwest::Client,
+    url: &str,
+) -> (RequestType, Result<Response, reqwest::Error>) {
+    let head_response = match client.head(url).send().await {
+        Ok(response) => response,
+        Err(e) => return (RequestType::Head, Err(e)),
+    };
+
+    // Successful head request
+    if head_response.status() != StatusCode::METHOD_NOT_ALLOWED {
+        return (RequestType::Head, Ok(head_response));
+    }
+
+    let result = client.get(url).send().await;
+    (RequestType::Get, result)
+}
+
+fn dns_error(err: &reqwest::Error) -> Option<String> {
+    let mut inner = &err as &dyn Error;
+    while let Some(source) = inner.source() {
+        inner = source;
+
+        // TODO: Would be better to get specific errors without string matching like this
+        // Not sure if there's a better way
+        let inner_message = inner.to_string();
+        if inner_message.contains("dns error") {
+            return Some(inner.source().unwrap().to_string());
+        }
     }
+    None
 }
 
 /// Makes a request to a url to determine whether it is up.
 /// Up is defined as returning a 2xx within a specific timeframe.
 pub async fn check_url(client: &reqwest::Client, url: &str) -> CheckResult {
-    match do_request(client, url).await {
-        Ok(_) => CheckResult::Success,
-        Err(e) => {
+    let trace_id = Uuid::new_v4();
+
+    let start = Instant::now();
+    let (request_type, response) = do_request(client, url).await;
+    let duration_ms = Some(start.elapsed().as_millis());
+
+    let status = if response.as_ref().map_or(false, |r| r.status().is_success()) {
```

Comment:
I just remembered that there is a nice new API for this that I've never used myself:
```suggestion
    let status = if response.is_ok_and(|r| r.status().is_success()) {
```

Reply:
ah perfect!

Reply:
Still wants the `as_ref()`

Reply:
Yeah that was my bad.